src/express-lane-auction/ExpressLaneAuction.sol

```diff
@@ -0,0 +1,538 @@
+        return _domainSeparatorV4();
+    }
+
+    /// @dev Internal bid domain hash
+    bytes32 private constant BID_DOMAIN =
+        keccak256("Bid(uint64 round,address expressLaneController,uint256 amount)");
+
+    /// @inheritdoc IExpressLaneAuction
+    function getBidHash(
+        uint64 round,
+        address expressLaneController,
+        uint256 amount
+    ) public view returns (bytes32) {
+        return
+            _hashTypedDataV4(
+                keccak256(abi.encode(BID_DOMAIN, round, expressLaneController, amount))
+            );
+    }
+
+    /// @notice Recover the signing address of the provided bid, and check that that address has enough funds to fulfil that bid
+    ///         Returns the signing address and the bid hash that was signed
+    /// @param bid The bid to recover the signing address of
+    /// @param biddingForRound The round the bid is for the control of
+    function recoverAndCheckBalance(
+        Bid memory bid,
+        uint64 biddingForRound,
+        RoundTimingInfo memory info
+    ) internal view returns (address, bytes32) {
+        bytes32 bidHash = getBidHash(biddingForRound, bid.expressLaneController, bid.amount);
+        address bidder = bidHash.recover(bid.signature);
+        // always check that the bidder has a much as they're claiming
+        if (_balanceOf[bidder].balanceAtRound(info.currentRound()) < bid.amount) {
+            revert InsufficientBalanceAcc(
+                bidder,
+                bid.amount,
+                _balanceOf[bidder].balanceAtRound(info.currentRound())
+            );
+        }
+
+        return (bidder, bidHash);
+    }
+
+    /// @inheritdoc IExpressLaneAuction
+    function resolveSingleBidAuction(Bid calldata firstPriceBid)
+        external
+        onlyRole(AUCTIONEER_ROLE)
+    {
+        RoundTimingInfo memory info = roundTimingInfo;
+        if (!info.isAuctionRoundClosed()) {
+            revert AuctionNotClosed();
+        }
+
+        if (firstPriceBid.amount < reservePrice) {
+            revert ReservePriceNotMet(firstPriceBid.amount, reservePrice);
+        }
+
+        uint64 biddingInRound = info.currentRound();
+        uint64 biddingForRound = biddingInRound + 1;
+        (address firstPriceBidder, ) = recoverAndCheckBalance(firstPriceBid, biddingForRound, info);
+
+        resolveAuction(false, firstPriceBid, firstPriceBidder, reservePrice, biddingInRound, info);
+    }
+
+    /// @inheritdoc IExpressLaneAuction
+    function resolveMultiBidAuction(Bid calldata firstPriceBid, Bid calldata secondPriceBid)
+        external
+        onlyRole(AUCTIONEER_ROLE)
+    {
+        RoundTimingInfo memory info = roundTimingInfo;
+        if (!info.isAuctionRoundClosed()) {
+            revert AuctionNotClosed();
+        }
+
+        // if the bids are the same amount and offchain mechanism will be used to choose the order and
+        // therefore the winner. The auctioneer is trusted to make this choice correctly
```

Comment:
the ordering is validated later instead of trusted

Reply:
removed comment